### ISSUE TITLE
Design support for multiple stdlibs

### ISSUE THREAD

**Issue by spalladino** [zeppelinos/zos-lib#218]
Design the changes required for supporting multiple dependencies (stdlib, package, etc) in a project. The goal is to open up the game for dev teams who want to push new features, available as zOS libs, without needing to reimplement or copy all the basic functionality already provided by OpenZeppelin-zOS.

**Comment by spalladino** [comment]
@facuspagnuolo @frangio I've just wrote this small doc on the ideas behind supporting multiple stdlibs. We should be able to tackle this in two stages: the first one with multiple dependencies, the second one with recursive dependencies. Let me know if you agree, and I can turn this into a task list for whoever tackles this.

---

Stdlibs (or packages) should be stored in the AppDirectoy as a mapping indexed by the package name, for contract lookup. Note that, when looking up contracts, we are currently referring to the stdlib contracts directly by name, same as if they were application contracts. This allows us to _shadow_ lib contracts using our own if needed. Having multiple stdlibs (or packages) may increase the risk of name clashing. Though one option is to define a lookup order, the safest option is to refer to the contracts with a fully qualified name. We can do this by either adding an optional `packageName` parameter to `AppDirectory#getImplementation`, or using a naming convention `packageName/contractName`. Js wrappers and the CLI need to be updated accordingly. Also, note that name clashing is not properly handled by solc/truffle, which could be an issue for development; but that issue is already present using regular off-chain libraries.

At the CLI level, packages should be stored in a `dependencies` or `packages` node (instad of a `stdlib` containing a single dependency), which contains the list of all packages installed in that application. The following commands would be affected:
- `link` should now add an stdlib instead of replacing the existing one
- `status` command should compare each stdlib individually
- `create` and `upgrade` should accept a fully-qualified name syntax when creating an instance for an stdlib
- `push` should connect to all stdlibs, or deploy all of them (TBD: should we allow some stdlibs to be uploaded by push, while others to be actually linked? should the user specify that, or should the CLI deploy for those stdlibs not available, and link for those that are available?)

Given that a dependency would now be able to provide a few specialised contracts, the need for recursive dependencies now arises. Though this could be handled in a future stage, we should nonetheless plan ahead. As we are mounting our package on top of npm, we already have recursive dependency management for free.

At the contracts level, a package with a dependency could have the same implementation as an `AppDirectory`. Furthermore, using fully-qualified names would ensure that requests for a package are correctly routed. For instance, if an app dependes on packages A and OZ, and A depends on OZ as well (albeit a different version), a request at the app for `OZ/ERC20` should always be routed to the direct dependency, and not to A's OZ. Note, again, that this clashing could pose problems at the solc/truffle level for compiling locally.

Custom deploys of packages (ie `push --deploy-stdlib`) would require crawling all packages dependencies, and deploying them as well. Their compiled artifacts should be available in the `node_modules` subfolder of each package. The `Contracts` contract provider would probably require some tuning to support this.

The `zos.network.json` could act fully as a package-lock now, as it should not only include the exact version of each package on that chain, but also the exact version of the recursive dependencies. Note that this would be purely informational, as the zOS `App` contract does not require any linking to the recursive dependencies, but only to the first-level ones.

**Comment by spalladino** [comment]
@facuspagnuolo should we close this one in favour of https://github.com/zeppelinos/zos-lib/issues/220?

**Comment by facuspagnuolo** [comment]
Thanks @spalladino for the great analysis! I do agree with the whole idea, let's start by supporting multiple stdlibs and then we can continue with recursive dependencies as you suggested. 

One little comment regarding how to point a particular implementation of a stdlib, I'd go with adding a new parameter to the `getImplementation` method, just to avoid doing extra string calculations in the contracts.

**Comment by spalladino** [comment]
> One little comment regarding how to point a particular implementation of a stdlib, I'd go with adding a new parameter to the getImplementation method, just to avoid doing extra string calculations in the contracts.

Great, I had already started using that approach ;-)

**Comment by spalladino** [comment]
After a long offline discussion with @frangio, we've decided on a few changes to the approach listed above. Without assigning a distinguished name to a `Package`, this would require duplicating all public functions that have a reference to a `contractName`. For instance, an app would require a `createProxy(contractName)` and a `createProxy(packageName, contractName)` method; the same would be repeated on several other functions.

The same problem appeared when designing for recursive dependencies. Each `Directory` would need to provide access to all of its dependencies, plus its own contracts, so a contract within it would be able to request a dependency. For instance, a `TokenWrapper` contract would need to be able to query its own `Directory` for an `OpenZeppelin/ERC20`. However, this `Directory` would also need to expose an inherent exported `Directory`, so when it is consumed by an app, it only exports its own contracts, and not its dependencies.

```solidity
contract Directory /* or Release of a Package */ {
  function getContract(namespace, contractName); // returns contracts from its own code or a dependency; to be called from its own contracts or an attached app
  function getContract(contractName); // returns contracts just from its code; to be used from another Package that uses it higher in the dependencies tree
}
```

This oddity appears because a Release has two responsibilities at the same time: 1) locking the versions of its dependencies, and providing access to them to its own contracts or an attacked lib, and 2) exposing a limited set of contracts to be used by another library that imports it (as the importing lib should not see the nested dependencies).

---

The agreement then was to lift the dependencies to the `App` itself for the time being. This way, an `App` has multiple `Package`s (which are either dependencies, or the ones that contain its own code) indexed by name, and pins a version for each `Package`. Each `Package` at a version returns a `Release` (formerly `AppDirectory`), which has a simple `getContract(name)` interface.

Later down the road, if we want to support recursive dependencies, a first approach could be to flatten all dependencies based on semver restrictions, and install all of them at the `App` level. This is opposed to the original approach of having nested recursive dependencies, that allowed multiple versions of the same dependency to coexist in the same tree.

--- 

The most hairy problem we encountered was how should we handle contracts creation within a library that depends on another one (again, let's say that `TokenWrapper` depends on `OZ/ERC20`). With the new approach, the library contract would need a reference to the `Release` corresponding to `OZ`, which should be injected by the `App` (since all dependencies are flattened and managed centrally by the App), and create a `Proxy` to it.

If we wanted to support nested recursive dependencies, as in the original approach, then the `TokenWrapper` implementation contract would need to have a fixed reference to either its `Release` or the `OZ` directory. This would require using a library-like linking system, as the reference must be included in the code itself, and not the storage. This is because the parent `App` cannot provide the reference to `OZ`, as it has no visibility to it, for it is managed directly by the `TokenWrapper` library.

---

Another topic for discussion was how an `App` would manage upgrading its dependencies. Today, an `App` is attached to a fixed `Package`, and can only move between this `Package` versions. However, it could be the case that, upon a fork, an `App` chooses to move to a different line of development (ie another `Package`). If so, we should contemplate for both 1) changing the version of a Package, and 2) changing the Package altogether. Most likely, these two methods would require different permissions, or different governance strategies.
